### ISSUE TITLE
Use `EpochScoring` in the default callbacks?

### ISSUE THREAD

**Issue by taketwo** [dnouri/skorch#220]
I noticed that `NeuralNet` and `NeuralNetClassifier` use `BatchScoring` as their default callbacks. In the documentation for the latter, however, it is explicitly recommended to use `EpochScoring` unless "you really need the scores for each batch".

**Comment by benjamin-work** [comment]
I'm glad that someone actually reads the docs ;)

`NeuralNet` et al. use `BatchScoring` for `train_loss` and `valid_loss`, but those losses are _special_. The reason is that they are determined during the training process anyway (since this is the loss that is optimized), so for those losses, we don't need an extra calculation, we just look up the values in the history (see [here](https://github.com/dnouri/skorch/blob/fc9d62657bc6d660e3d63e9e6745301811e7dfca/skorch/net.py#L34)). By using `BatchScoring`, we make sure that the losses are averaged correctly over the batches.

In short: Since we already have the "scores for each batch", it makes sense to use them.

Now this part is not documented further because usually, the user doesn't need to bother with it.

**Comment by taketwo** [comment]
The skorch docs are very informative and are a great help in getting started!

Thanks for the explanation. Now if I understand correctly, `EpochScoring` will also avoid inference thanks to caching. Though it will still need to recompute losses based on the cached output, hence `BatchScoring` is preferable.

> Now this part is not documented further because usually, the user doesn't need to bother with it.

I made a noob mistake of not shuffling my classification dataset and had to go through all the skorch internals with a debugger to understand why the train loss goes down, but validation accuracy stays the same.
